### DISCUSSION TITLE
When passing an async server action to useActionState, the returned dispatch is not a Promise.

### DISCUSSION THREAD

**Discussion by mrbjjackson** [vercel/next.js#67124] [Help]
### Summary

I'm trying out the latest Next 15 RC as I'd love to use useActionState in my app.
I have an server action which works fine as an asynchronous function - you can "wait" for the response. However if I pass the function to useActionState and use the generated dispatch function, it does not work as an async function.

Is this intentional? Is it a bug? I can't see a reason why this should be the case.

In the example code below, the `demoFunction()` behaves as a Promise fine. But when I run it through `useActionState`, the generated `demoAction()` function does not. Obviously I can use the `demoFunction `version but then I don't get the "isPending" helper.

```
....

async function demoFunction() {
  'use server'
  await new Promise((resolve) => setTimeout(resolve, 3000));
  return 'success'
}

const [state, demoAction, isPending] = useActionState(demoFunction,   null );


const handleClickOne = async () => {
   await demoFunction()
   console.log('FIRES THREE SECONDS AFTER CLICK')
}

const handleClickTwo = async () => {
   await demoAction()
   console.log('FIRES IMMEDIATELY')
}

....

<Button onClick={handleClickOne}>Not Using useActionState</Button>

<Button onClick={handleClickTwo}>Using useActionState</Button>

```

**Comment by icyJoseph** [comment]
This sounds like it belongs to React's repository rather, https://react.dev/reference/react/useActionState#noun-labs-1201738-(2):

> A new action that you can pass as the action prop to your form component or formAction prop to any button component within the form.

This is stated a couple of times in their docs, and as I interpret it, the action you get from the hook, is a trigger, for a bunch of work, that includes, firing the action you passed, and collecting its result to assign it as the next state.

**Reply by chelorossi** [reply]
@mrbjjackson  I think you need to use startTransition for server actions to grabbing the isPending right, check this out: [https://react.dev/reference/rsc/server-actions](https://react.dev/reference/rsc/server-actions).
You are composing client with server actions so: https://react.dev/reference/rsc/server-actions#composing-server-actions-with-actions
